lib/Sema/TypeCheckStmt.cpp

```diff
@@ -352,13 +355,16 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
   };
 
   StmtChecker(TypeChecker &TC, AbstractFunctionDecl *AFD)
-    : TC(TC), TheFunc(AFD), DC(AFD), IsREPL(false) { }
+      : TC(TC), TheFunc(AFD), DC(AFD), IsREPL(false),
+        IsBraceStmtFromTopLevelDecl(false) {}
 
   StmtChecker(TypeChecker &TC, ClosureExpr *TheClosure)
-    : TC(TC), TheFunc(TheClosure), DC(TheClosure), IsREPL(false) { }
+      : TC(TC), TheFunc(TheClosure), DC(TheClosure), IsREPL(false),
+        IsBraceStmtFromTopLevelDecl(false) {}
 
   StmtChecker(TypeChecker &TC, DeclContext *DC)
-    : TC(TC), TheFunc(), DC(DC), IsREPL(false) {
+      : TC(TC), TheFunc(), DC(DC), IsREPL(false),
+        IsBraceStmtFromTopLevelDecl(true) {
```

Comment:
I missed this earlier. Rather than setting this to `true`, can you instead check `getContextKind()` of the `DeclContext` that is passed in. That is less fragile, since as written this relies on the fact that the only place we currently use this constructor is when we're passing in a top level decl context.

Reply:
I think this is kind of similar to what Robert suggested above about using `isa<TopLevelCodeDecl>(DC)`,  to what I responded:

> Unfortunately I can not relay on `isa<TopLevelCodeDecl>(DC)` because in the following code snippet, it would return `true` when visiting the `{` that opens the `if`s `then` block:

```swift
let x = 1
let y = 3

if (x > y) {
    defer {
        print("not so useful defer stmt.")
    }
}
```
> and I would like to diagnose this block.

when visiting the open brace of the if `DC->getContextKind()` returns `TopLevelCodeDecl`.

Reply:
For that you should have an outer `BraceStmt`, with an `IfStmt` nested within it, and a `BraceStmt` nested under it. So I would expect that when you hit the outermost `BraceStmt` you would set `IsBraceStmtFromTopLevelDecl` to false like you are now.

Reply:
sorry.. I completely missed your point. I see what you mean now, I should base the initialization of the IsBraceStmtFromTopLevelDecl on `DC->getContextKind()`. Fixed it on da3149d713f5c9e7b1faed096482c66e7c70d0e4